File: ios/HybridBackgroundRenderer.swift
//
//  HybridBackgroundRenderer.swift
//  VisionCameraBackgroundFilter
//

import AVFoundation
import CoreImage
import CoreVideo
import Foundation
import Metal
import NitroModules
import VisionCamera

/**
 * The live compositor.
 *
 * `renderFrame` is called from the `CameraFrameOutput` thread. The CoreImage
 * context and the Metal command queue are created once and reused — building a
 * `CIContext` per frame is the classic way to turn a 30fps preview into a 5fps one.
 */
final class HybridBackgroundRenderer: HybridBackgroundRendererSpec, NativeSurfaceRenderer {

  private let segmenter = SelfieSegmenter(quality: .balanced)
  private let compositor = BackgroundCompositor()

  private let metalDevice: MTLDevice?
  private let commandQueue: MTLCommandQueue?
  private let ciContext: CIContext?

  /// The layer to present into, handed over by the view.
  private var targetLayer: CAMetalLayer?

  var isSupported: Bool {
    return SelfieSegmenter.isSupported && metalDevice != nil
  }

  var unsupportedReason: String? {
    if let reason = SelfieSegmenter.unsupportedReason {
      return reason.rawValue
    }
    // Every device that runs iOS 15 has Metal, so this is defensive only.
    return metalDevice == nil ? SegmentationUnsupportedReason.unsupportedDevice.rawValue : nil
  }

  override init() {
    let device = MTLCreateSystemDefaultDevice()
    metalDevice = device
    commandQueue = device?.makeCommandQueue()
    if let device {
      // `.workingColorSpace: NSNull` disables CoreImage's colour management for the
      // blend. It is both faster and correct here: the camera frame and the
      // background are already in display space, so a conversion would shift hues.
      ciContext = CIContext(
        mtlDevice: device,
        options: [.workingColorSpace: NSNull(), .cacheIntermediates: false]
      )
    } else {
      ciContext = nil
    }
    super.init()
  }

  /// See `setCameraMirrored`. Written from JS, read on the frame thread.
  private var cameraMirrored = false

  /// The in-flight record-time capture, if any.
  ///
  /// Written from the JS thread (`startCapture` / `stopCapture`), read and fed from
  /// the frame thread — hence the lock. It is held across the whole append so a stop
  /// cannot free the writer out from under a frame already being encoded; the
  /// critical section is one `CIContext.render`, which the frame thread was going to
  /// pay for anyway.
  private var recorder: CompositeRecorder?
  private let recorderLock = NSLock()

  func setBackground(uri: String?, fit: BackgroundFit, mirror: Bool) throws {
    let mode: BackgroundFitMode = fit == .cover ? .cover : .contain
    guard let uri, let image = BackgroundImageLoader.load(uri: uri) else {
      compositor.setBackground(nil, fit: mode, mirror: mirror)
      return
    }
    compositor.setBackground(image, fit: mode, mirror: mirror)
  }

  func setCameraMirrored(mirrored: Bool) throws {
    cameraMirrored = mirrored
  }

  func renderFrame(frame: any HybridFrameSpec) throws {
    guard let layer = targetLayer,
          let ciContext,
          let commandQueue
    else { return }

    guard let nativeFrame = frame as? any NativeFrame else {
      throw RuntimeError.error(withMessage: "Frame is not of type `NativeFrame`!")
    }
    guard let sampleBuffer = nativeFrame.sampleBuffer,
          let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer)
    else {
      throw RuntimeError.error(withMessage: "Frame does not have a valid pixel buffer!")
    }

    // Camera buffers arrive in SENSOR orientation, not display orientation — on a
    // portrait iPhone the sensor is landscape, so drawing the buffer as-is renders
    // the preview rotated 90°; it is the consumer's job to counter-rotate, and not
    // doing so was the "flipped 90° to the left" bug.
    //
    // `frame.orientation` is what to counter-rotate BY, but it does not describe the
    // buffer — this comment used to claim it did, and that premise is what left the
    // rear camera upside down for a release. It is the connection's output
    // convention, and it differs by 180° between the two cameras; see
    // `cgOrientation(for:isFrontCamera:)`. `cameraMirrored` doubles as "the front
    // camera is open", which is what the consumer is asked to report (see the spec
    // for `setCameraMirrored`), so it is what the rear camera is normalised against.
    let cgOrientation = Self.cgOrientation(
      for: frame.orientation,
      isFrontCamera: cameraMirrored
    )
    let source = CIImage(cvPixelBuffer: pixelBuffer).oriented(cgOrientation)

    // TEMPORARY diagnostic — remove before merging. The rear camera's filtered
    // preview comes out rotated 180° against the unfiltered one while the front
    // camera is correct, and the same code path serves both, so the only input that
    // can differ is what arrives here. Logs once per change, not per frame.
    Self.logOrientationOnce(
      orientation: frame.orientation,
      isMirrored: frame.isMirrored,
      cameraMirrored: cameraMirrored,
      cgOrientation: cgOrientation,
      rawExtent: CIImage(cvPixelBuffer: pixelBuffer).extent,
      orientedExtent: source.extent
    )

    // The SAME orientation goes to Vision, so the mask comes back in the corrected
    // space and lines up with `source`. Passing `.up` here (as this did) also gave
    // the model a sideways person to segment.
    let mask = compositor.hasBackground
      ? segmenter.segment(pixelBuffer: pixelBuffer, orientation: cgOrientation)
      : nil
    // XOR, so the two mirroring sources cannot cancel into a double flip: the
    // consumer asks for a mirrored *presentation*, while `isMirrored` reports that
    // the buffer's own handedness already differs from the output's. Measured on
    // Android, `isMirrored` is false even on a selfie camera whose preview is
    // mirrored, which is why the consumer has to tell us (see the spec).
    let mirrorFrame = cameraMirrored != frame.isMirrored
    let output = compositor.composite(frame: source, mask: mask, mirrorFrame: mirrorFrame)

    // Encode BEFORE presenting: the recorded file is the deliverable, and a frame
    // the encoder was not ready for is lost, whereas a preview frame arriving a
    // millisecond later is not noticeable.
    //
    // The capture composites UNMIRRORED even when the preview is mirrored — exactly
    // as the offline bake and the `expo-camera` path do, so a recording made with
    // the filter matches one made without it. That costs a second `CIBlendWithMask`
    // on the front camera; the expensive part, segmentation, still runs once.
    recorderLock.lock()
    if let recorder {
      let captureImage = mirrorFrame
        ? compositor.composite(frame: source, mask: mask, mirrorFrame: false)
        : output
      recorder.append(
        captureImage,
        at: CMSampleBufferGetPresentationTimeStamp(sampleBuffer),
        using: ciContext
      )
    }
    recorderLock.unlock()

    // Keep the drawable sized to the frame so CoreImage never rescales twice.
    layer.drawableSize = CGSize(width: output.extent.width, height: output.extent.height)
    guard let drawable = layer.nextDrawable(),
          let commandBuffer = commandQueue.makeCommandBuffer()
    else { return }

    let destination = CIRenderDestination(
      width: Int(layer.drawableSize.width),
      height: Int(layer.drawableSize.height),
      pixelFormat: layer.pixelFormat,
      commandBuffer: commandBuffer,
      mtlTextureProvider: { drawable.texture }
    )
    // CoreImage is bottom-left origin, the drawable is top-left; without this the
    // preview would be upside down.
    destination.isFlipped = true

    _ = try? ciContext.startTask(toRender: output, to: destination)
    commandBuffer.present(drawable)
    commandBuffer.commit()
  }

  // MARK: - Record-time capture

  /// `AVAssetWriter` is available on every supported OS; what this really reports is
  /// whether the CoreImage context that renders into it exists.
  var isCaptureSupported: Bool {
    return ciContext != nil
  }

  func startCapture(outputPath: String, maxOutputHeight: Double, aspectRatio: Double) throws {
    guard isCaptureSupported else {
      throw RuntimeError.error(withMessage: "This device cannot encode composited frames.")
    }
    recorderLock.lock()
    defer { recorderLock.unlock() }
    guard recorder == nil else {
      throw RuntimeError.error(withMessage: "A capture is already running.")
    }
    recorder = CompositeRecorder(
      outputPath: outputPath,
      maxOutputHeight: Int(maxOutputHeight),
      aspectRatio: aspectRatio
    )
  }

  func stopCapture() throws -> Promise<String> {
    recorderLock.lock()
    let running = recorder
    recorder = nil
    recorderLock.unlock()

    let promise = Promise<String>()
    guard let running else {
      throw RuntimeError.error(withMessage: "No capture is running.")
    }
    // `finishWriting` is asynchronous and its completion arrives on an AVFoundation
    // queue; the promise is resolved from there rather than blocking the caller.
    running.finish { result in
      switch result {
      case .success(let path):
        promise.resolve(withResult: path)
      case .failure(let error):
        promise.reject(withError: error)
      }
    }
    return promise
  }

  deinit {
    // A screen torn down mid-recording would otherwise leave a writer holding an
    // unfinished file that nothing else knows the path of.
    recorderLock.lock()
    recorder?.cancel()
    recorder = nil
    recorderLock.unlock()
  }

  // MARK: - NativeSurfaceRenderer

  func connectLayer(_ layer: CAMetalLayer) {
    layer.device = metalDevice
    layer.pixelFormat = .bgra8Unorm
    layer.framebufferOnly = false
    // The drawable is sized to the camera frame, but the layer is laid out to the
    // view — and `contentsGravity` defaults to `.resize`, which STRETCHES one onto
    // the other. A 9:16 frame in a view of any other aspect then comes out visibly
    // too wide or too tall, which is not what the unfiltered preview does.
    // `.resizeAspectFill` centre-crops instead, matching VisionCamera's own
    // PreviewView (its `resizeMode` defaults to 'cover').
    layer.contentsGravity = .resizeAspectFill
    targetLayer = layer
  }

  func disconnectLayer() {
    targetLayer = nil
  }

  // MARK: - Background decoding

  /**
   * Maps VisionCamera's `CameraOrientation` to the `CGImagePropertyOrientation`
   * that corrects the delivered buffer.
   *
   * The names line up one-to-one with VisionCamera's own mapping in
   * `CG+CameraOrientation.swift`, reimplemented because that extension is internal
   * to the VisionCamera module — but the value it is given is normalised first, which
   * VisionCamera's version does not do. See the body.
   *
   * Mirroring is deliberately NOT folded in here (no `.leftMirrored` etc.) — it is
   * applied in the compositor instead, so the mask is flipped with the frame. That
   * decomposition is exact rather than approximate: `flipH ∘ rot90CCW` is the
   * transpose, i.e. `.leftMirrored`, and `flipH ∘ rot90CW` is the transverse, i.e.
   * `.rightMirrored`. Folding mirroring in would change nothing.
   *
   * ⚠️ This used to claim "the front camera's buffer already arrives mirrored, so
   * leaving it alone keeps the preview mirror-like". That assumption was **wrong on
   * Android** — the frame output's buffer is not mirrored and `Frame.isMirrored` is
   * `false`, so the filtered preview came out as a mirror image of the unfiltered
   * one. It is unverified on iOS, and is now handled explicitly either way via
   * `setCameraMirrored`. The background stays unmirrored in both the live and offline
   * paths, so only the person is mirrored — as on the `expo-camera` path.
   */
  private static func cgOrientation(
    for orientation: CameraOrientation,
    isFrontCamera: Bool
  ) -> CGImagePropertyOrientation {
    // ─── Why the rear camera is turned through 180° ───
    //
    // `frame.orientation` is `AVCaptureConnection.videoOrientation` read back
    // (VisionCamera's `AVCaptureConnection+orientation.swift`). That is the
    // connection's OUTPUT convention, not a description of the buffer that was
    // delivered — and AVFoundation's default differs by 180° between the two
    // cameras, because the front camera's convention has its mirrored mount baked
    // in. Measured on a portrait iPhone, same device orientation, both cameras:
    //
    //   front: orientation .right, isMirrored false, buffer 1920x1080
    //   rear:  orientation .left,  isMirrored false, buffer 1920x1080
    //
    // Identical buffers, values 180° apart. So they cannot both be counter-rotated
    // by their own reported value: the front comes out upright and the rear upside
    // down, which is exactly the reported defect. The buffers agree, so the
    // correction must too — normalise onto the front camera's value, which is the
    // one that renders upright.
    //
    // Adding 180° (rather than inverting) is what closes the measured gap: `.left`
    // and `.right` are 180° apart, as are `.up` and `.down`, so this generalises to
    // a rotating host instead of only holding in portrait.
    let corrected = isFrontCamera ? orientation : turnedThrough180(orientation)
    switch corrected {
    case .up: return .up
    case .down: return .down
    case .left: return .left
    case .right: return .right
    default: return .up
    }
  }

  /// `CameraOrientation` turned through 180°: `.left`/`.right` are opposites, and so
  /// are `.up`/`.down`.
  private static func turnedThrough180(_ orientation: CameraOrientation) -> CameraOrientation {
    switch orientation {
    case .up: return .down
    case .down: return .up
    case .left: return .right
    case .right: return .left
    default: return orientation
    }
  }

  /// TEMPORARY diagnostic — remove before merging. See the call site in `renderFrame`.
  private static var lastOrientationSignature: String?
  private static func logOrientationOnce(
    orientation: CameraOrientation,
    isMirrored: Bool,
    cameraMirrored: Bool,
    cgOrientation: CGImagePropertyOrientation,
    rawExtent: CGRect,
    orientedExtent: CGRect
  ) {
    let signature = """
    [BackgroundFilter] frame.orientation=\(orientation) frame.isMirrored=\(isMirrored) \
    cameraMirrored=\(cameraMirrored) -> cgOrientation=\(cgOrientation.rawValue) \
    (expect 6 on BOTH cameras in portrait) \
    mirrorFrame=\(cameraMirrored != isMirrored) \
    raw=\(Int(rawExtent.width))x\(Int(rawExtent.height)) \
    oriented=\(Int(orientedExtent.width))x\(Int(orientedExtent.height))
    """
    guard signature != lastOrientationSignature else { return }
    lastOrientationSignature = signature
    NSLog("%@", signature)
  }

}

/// Contract between the renderer and the view that hosts its output.
protocol NativeSurfaceRenderer: AnyObject {
  func connectLayer(_ layer: CAMetalLayer)
  func disconnectLayer()
}

File: android/src/main/java/com/margelo/nitro/backgroundfilter/HybridBackgroundRenderer.kt
package com.margelo.nitro.backgroundfilter

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.Image
import android.view.Surface
import androidx.annotation.OptIn
import androidx.camera.core.ExperimentalGetImage
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.`public`.NativeFrame
import com.margelo.nitro.core.Promise
import com.saigontechnology.backgroundfilter.BackgroundFitMode
import com.saigontechnology.backgroundfilter.capture.CompositeVideoRecorder
import com.saigontechnology.backgroundfilter.BackgroundGeometry
import com.saigontechnology.backgroundfilter.NativeSurfaceRenderer
import com.saigontechnology.backgroundfilter.SegmentationQuality
import com.saigontechnology.backgroundfilter.SelfieSegmenter
import com.saigontechnology.backgroundfilter.gl.CompositeGl
import com.saigontechnology.backgroundfilter.gl.EglCore
import java.io.File
import java.net.URL
import java.nio.ByteBuffer
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * The live compositor.
 *
 * `renderFrame` runs on the `CameraFrameOutput` thread, which is also the thread
 * that owns the EGL context and every GL object. Nothing here is synchronized:
 * thread confinement is the safety argument, and the only cross-thread state is
 * [pendingBackground], which is written from JS and read on the render thread —
 * hence `@Volatile`.
 *
 * This class lives in `com.margelo.nitro.backgroundfilter` because nitrogen
 * hard-codes `com.margelo.nitro` as the base package for generated specs and
 * their implementations. Our own non-Nitro code stays under
 * `com.saigontechnology.backgroundfilter`.
 */
@OptIn(ExperimentalGetImage::class)
class HybridBackgroundRenderer :
  HybridBackgroundRendererSpec(),
  NativeSurfaceRenderer {

  private companion object {
    /**
     * Mirrors `MAX_BACKGROUND_EDGE_PX` in `src/assets/background/index.ts`.
     * Duplicated because native cannot read the TS constant; keep them in step.
     */
    const val MAX_BACKGROUND_EDGE_PX = 1920
  }

  private val segmenter = SelfieSegmenter(SegmentationQuality.BALANCED)
  private val egl = EglCore()
  private val composite = CompositeGl()

  private var isGlReady = false
  private var viewportWidth = 0
  private var viewportHeight = 0

  /** Set from JS, consumed on the render thread on the next frame. */
  @Volatile
  private var pendingBackground: PendingBackground? = null

  private var backgroundBitmap: Bitmap? = null
  private var fit: BackgroundFitMode = BackgroundFitMode.COVER
  private var mirror: Boolean = false

  /** Reused across frames so a 1080p mask isn't reallocated 30 times a second. */
  private var maskBytes: ByteBuffer? = null

  /** See `setCameraMirrored`. Plain field: read on the render thread, set from JS. */
  @Volatile
  private var cameraMirrored: Boolean = false

  private data class PendingBackground(val uri: String?, val fit: BackgroundFitMode, val mirror: Boolean)

  // ─── Record-time capture ───────────────────────────────────────────────────
  //
  // The encoder is owned by the GL thread (its input surface lives in that thread's
  // EGL context), but start/stop arrive from JS. So JS only ever posts a REQUEST,
  // which the next frame picks up, and everything that touches [recorder] runs
  // inside [captureLock] — including the per-frame draw, so a stop cannot free the
  // encoder out from under a frame already being drawn into it.

  private data class PendingCapture(
    val outputPath: String,
    val maxOutputHeight: Int,
    val aspectRatio: Double,
  )

  private val captureLock = Any()

  @Volatile
  private var pendingCapture: PendingCapture? = null

  private var recorder: CompositeVideoRecorder? = null

  /** Timestamp of the capture's first frame; every later one is written relative to it. */
  private var captureStartNanos = 0L

  /** Set when the encoder could not be configured; surfaced by `stopCapture`. */
  @Volatile
  private var captureFailure: Throwable? = null

  /**
   * Finishing the encoder blocks — it drains to end-of-stream and writes the moov
   * atom — so it runs off the JS thread. Single-threaded: only one capture can be
   * in flight, and finishing them in order is what the promise contract implies.
   */
  private val captureFinisher: ExecutorService =
    Executors.newSingleThreadExecutor { runnable ->
      Thread(runnable, "BackgroundFilterCapture").apply { isDaemon = true }
    }

  override val isSupported: Boolean
    get() = SelfieSegmenter.isSupported

  override val unsupportedReason: String?
    get() = SelfieSegmenter.unsupportedReason?.value

  override fun setBackground(uri: String?, fit: BackgroundFit, mirror: Boolean) {
    val mode = when (fit) {
      BackgroundFit.COVER -> BackgroundFitMode.COVER
      BackgroundFit.CONTAIN -> BackgroundFitMode.CONTAIN
    }
    // Decoding and texture upload happen on the render thread; this only records
    // the request so JS never blocks on IO.
    pendingBackground = PendingBackground(uri, mode, mirror)
  }

  override fun setCameraMirrored(mirrored: Boolean) {
    cameraMirrored = mirrored
  }

  /**
   * Every device with a camera has a hardware H.264 encoder, so this tracks the
   * segmentation gate: without a mask there is no composite worth recording, and the
   * consumer would be better off using the camera's own recorder.
   */
  override val isCaptureSupported: Boolean
    get() = SelfieSegmenter.isSupported

  override fun startCapture(outputPath: String, maxOutputHeight: Double, aspectRatio: Double) {
    synchronized(captureLock) {
      if (recorder != null || pendingCapture != null) {
        throw IllegalStateException("A capture is already running.")
      }
      captureFailure = null
      captureStartNanos = 0L
      pendingCapture = PendingCapture(outputPath, maxOutputHeight.toInt(), aspectRatio)
    }
  }

  override fun stopCapture(): Promise<String> {
    val promise = Promise<String>()
    val running: CompositeVideoRecorder?
    val failure: Throwable?
    synchronized(captureLock) {
      running = recorder
      failure = captureFailure
      recorder = null
      pendingCapture = null
      captureFailure = null
    }
    // Clearing [recorder] above is what makes the rest of this safe: the frame thread
    // checks it under the same lock and stops touching the encoder immediately.

    when {
      failure != null -> promise.reject(failure)
      running == null -> promise.reject(IllegalStateException("No capture is running."))
      else -> captureFinisher.execute {
        try {
          promise.resolve(running.finish())
        } catch (e: Throwable) {
          promise.reject(e)
        } finally {
          // AFTER the encoder is done, never before: `finish()` signals end-of-stream
          // on the input surface and drains what comes back, and tearing the surface
          // down first is how that turns into a hang or an IllegalStateException.
          // Destroying it from this thread is fine — EGL defers the destroy while a
          // surface is current elsewhere.
          runCatching { egl.releaseEncoderSurface() }
        }
      }
    }
    return promise
  }

  override fun renderFrame(frame: HybridFrameSpec) {
    if (!egl.hasSurface) return

    val nativeFrame = frame as? NativeFrame ?: throw Error("Frame is not of type `NativeFrame`!")
    val image = nativeFrame.image.image ?: return

    // This runs on the frame-delivery thread, NOT the thread that created the
    // window surface — so the context has to be bound here. Without it every GL
    // call below is a silent no-op (see the note in EglCore). A failed bind means
    // the surface went away underneath us (the screen is being torn down), so drop
    // the frame rather than drawing into nothing.
    if (!egl.makeCurrent()) return

    applyPendingBackground()

    if (!isGlReady) {
      composite.setUp()
      isGlReady = true
    }
    uploadPendingBitmap()

    val rotation = nativeFrame.image.imageInfo.rotationDegrees
    val mask = if (composite.isBackgroundUploaded) segment(image, rotation) else null

    // A 90/270 rotation swaps the frame's DISPLAYED dimensions.
    val quarterTurned = rotation == 90 || rotation == 270
    val displayWidth = if (quarterTurned) image.height else image.width
    val displayHeight = if (quarterTurned) image.width else image.height

    // Everything is composed in the VIEWPORT's aspect, not the frame's. The camera
    // frame is cover-cropped into the viewport exactly as VisionCamera's own
    // PreviewView does (its `resizeMode` defaults to 'cover'), so turning the filter
    // on does not change the subject's shape. Sampling the full frame instead
    // stretched it onto whatever aspect the view had — a 9:16 frame in a 0.71 view
    // came out ~27% too wide.
    val cameraLayout = BackgroundGeometry.computeBackgroundLayout(
      bgWidth = displayWidth.toFloat(),
      bgHeight = displayHeight.toFloat(),
      frameWidth = viewportWidth.toFloat(),
      frameHeight = viewportHeight.toFloat(),
      fit = BackgroundFitMode.COVER,
    )
    val frameUv = BackgroundGeometry.sourceToTextureCoords(
      cameraLayout,
      displayWidth.toFloat(),
      displayHeight.toFloat(),
      // Never mirror via the rect: a centred crop is symmetric, so flipping it is a
      // no-op. Mirroring is applied by swapping the U bounds below.
      false,
    )
    // `Frame.isMirrored` means "the output is mirrored but these pixels are not", so
    // a renderer has to flip X itself — VisionCamera's docs say exactly that. With
    // the default `mirrorMode: 'auto'` this is true on the selfie camera, and
    // ignoring it made the filtered preview a mirror image of the unfiltered one.
    // XOR, so the two mirroring sources cannot cancel into a double flip: the
    // consumer asks for a mirrored *presentation*, while `isMirrored` reports that
    // the buffer's own handedness already differs from the output's.
    if (cameraMirrored != frame.isMirrored) {
      val minU = frameUv[0]
      frameUv[0] = frameUv[2]
      frameUv[2] = minU
    }

    // The background is fitted to the viewport for the same reason as the camera —
    // it is drawn over the same quad, so fitting it to the frame would stretch it by
    // the frame-to-viewport aspect difference.
    val layout = BackgroundGeometry.computeBackgroundLayout(
      bgWidth = (backgroundBitmap?.width ?: 0).toFloat(),
      bgHeight = (backgroundBitmap?.height ?: 0).toFloat(),
      frameWidth = viewportWidth.toFloat(),
      frameHeight = viewportHeight.toFloat(),
      fit = fit,
    )
    val backgroundUv = BackgroundGeometry.sourceToTextureCoords(
      layout,
      (backgroundBitmap?.width ?: 1).toFloat(),
      (backgroundBitmap?.height ?: 1).toFloat(),
      mirror,
    )

    val planes = image.planes
    composite.uploadFrame(
      yPlane = planes[0].buffer,
      uvPlane = planes[1].buffer,
      frameWidth = image.width,
      frameHeight = image.height,
      mask = mask?.buffer,
      maskWidth = mask?.width ?: 0,
      maskHeight = mask?.height ?: 0,
      // Rows are padded to a hardware stride that is not always the width — a
      // 1280x960 buffer on a Galaxy S22 is padded where a 960x720 one is not.
      // Ignoring it shears the image into vertical stripes.
      yRowStride = planes[0].rowStride,
      uvRowStride = planes[1].rowStride,
    )
    composite.drawUploaded(
      backgroundUv = backgroundUv,
      frameUv = frameUv,
      viewportWidth = viewportWidth,
      viewportHeight = viewportHeight,
      rotationDegrees = rotation,
    )
    egl.swapBuffers()

    captureFrame(displayWidth, displayHeight, rotation, nativeFrame.image.imageInfo.timestamp)
  }

  /**
   * Draws the frame just uploaded a second time, into the capture encoder.
   *
   * Deliberately a separate draw rather than a copy of the preview: the two targets
   * want different geometry. The preview is cover-cropped into whatever aspect the
   * view has; the recording is cover-cropped into the ENCODER's aspect, which the
   * consumer sets to whatever the camera's own recorder writes. Both are unmirrored,
   * as the offline bake and the `expo-camera` path are, so a clip made with the
   * filter matches one made without it in shape as well as handedness.
   *
   * The textures are already resident, so this costs one draw call, not a second
   * upload.
   */
  private fun captureFrame(
    displayWidth: Int,
    displayHeight: Int,
    rotation: Int,
    timestampNanos: Long,
  ) {
    synchronized(captureLock) {
      startPendingCapture(displayWidth, displayHeight)
      val recorder = this.recorder ?: return
      if (!egl.makeCurrentEncoder()) return

      val layout = BackgroundGeometry.computeBackgroundLayout(
        bgWidth = (backgroundBitmap?.width ?: 0).toFloat(),
        bgHeight = (backgroundBitmap?.height ?: 0).toFloat(),
        frameWidth = recorder.width.toFloat(),
        frameHeight = recorder.height.toFloat(),
        fit = fit,
      )
      val backgroundUv = BackgroundGeometry.sourceToTextureCoords(
        layout,
        (backgroundBitmap?.width ?: 1).toFloat(),
        (backgroundBitmap?.height ?: 1).toFloat(),
        mirror,
      )

      // Cover-crop the camera into the encoder's aspect, exactly as the preview pass
      // crops it into the view's. `FULL_FRAME_UV` was right only while the two shapes
      // matched; now the frames are 4:3 (to match the preview's field of view) and the
      // recording is the recorder's shape, so sampling the whole frame would squash it.
      val cameraLayout = BackgroundGeometry.computeBackgroundLayout(
        bgWidth = displayWidth.toFloat(),
        bgHeight = displayHeight.toFloat(),
        frameWidth = recorder.width.toFloat(),
        frameHeight = recorder.height.toFloat(),
        fit = BackgroundFitMode.COVER,
      )
      val captureFrameUv = BackgroundGeometry.sourceToTextureCoords(
        cameraLayout,
        displayWidth.toFloat(),
        displayHeight.toFloat(),
        // Never mirrored: recorded files are not, on any path.
        false,
      )

      composite.drawUploaded(
        backgroundUv = backgroundUv,
        frameUv = captureFrameUv,
        viewportWidth = recorder.width,
        viewportHeight = recorder.height,
        rotationDegrees = rotation,
      )
      // The timestamp goes on the surface, not the buffer, and has to be stamped
      // between the draw and the swap. Without it MediaCodec times frames by when
      // the swap happened, which drifts against the audio track we mux in later.
      //
      // Camera timestamps are absolute (nanos since boot), and neither MediaCodec
      // nor MediaMuxer rebases them — a clip whose first frame is at t=12345s is
      // what that produces. So the first captured frame defines zero.
      if (captureStartNanos == 0L) captureStartNanos = timestampNanos
      egl.setEncoderPresentationTime((timestampNanos - captureStartNanos).coerceAtLeast(0L))
      egl.swapEncoder()
      recorder.drain(endOfStream = false)
    }
  }

  /**
   * Configures the encoder on the first frame after `startCapture`.
   *
   * Deferred to the GL thread on purpose: the encoder's input surface has to be
   * wrapped in an `EGLSurface` on the thread that owns the context, and the output
   * size is not known until a frame arrives with its displayed dimensions.
   *
   * A configuration failure disables capture for this run rather than throwing into
   * the frame pipeline — the preview keeps working, and `stopCapture` reports it.
   */
  private fun startPendingCapture(displayWidth: Int, displayHeight: Int) {
    val request = pendingCapture ?: return
    pendingCapture = null

    val recorder =
      CompositeVideoRecorder(request.outputPath, request.maxOutputHeight, request.aspectRatio)
    try {
      val surface = recorder.start(displayWidth, displayHeight)
      egl.createEncoderSurface(surface)
      this.recorder = recorder
    } catch (e: Throwable) {
      runCatching { recorder.release(deleteOutput = true) }
      captureFailure = e
    }
  }

  private class Mask(val buffer: ByteBuffer, val width: Int, val height: Int)

  /**
   * Segments and converts MLKit's float confidences to the 8-bit texture the
   * shader samples. Returns null when the frame produced no mask — the caller then
   * draws the frame unmodified rather than reusing a stale silhouette.
   */
  private fun segment(image: Image, rotationDegrees: Int): Mask? {
    val result = segmenter.segment(image, rotationDegrees) ?: return null
    val floats = result.buffer
    floats.rewind()
    val pixelCount = result.width * result.height
    val target = maskBytes?.takeIf { it.capacity() >= pixelCount }
      ?: ByteBuffer.allocateDirect(pixelCount).also { maskBytes = it }
    target.clear()
    for (i in 0 until pixelCount) {
      target.put((floats.float.coerceIn(0f, 1f) * 255f).toInt().toByte())
    }
    target.rewind()
    floats.rewind()
    return Mask(target, result.width, result.height)
  }

  private fun applyPendingBackground() {
    val request = pendingBackground ?: return
    pendingBackground = null
    fit = request.fit
    mirror = request.mirror

    backgroundBitmap?.recycle()
    backgroundBitmap = request.uri?.let { decode(it) }
    // ALWAYS invalidate the uploaded texture, not just when the new background is
    // null. The texture on the GPU belongs to the bitmap we just recycled, and
    // [uploadPendingBitmap] skips the upload while `isBackgroundUploaded` is set —
    // so clearing it only for `null` meant switching from one background straight to
    // another kept showing the first. Going via "none" appeared to be required,
    // because that was the only path that cleared the flag.
    composite.clearBackground()
  }

  /**
   * Uploads the current background's texture, once per change.
   *
   * The guard is what keeps this off the per-frame path: `isBackgroundUploaded` is
   * cleared by [applyPendingBackground] whenever the bitmap changes and by
   * [connectSurface] when the GL context is rebuilt, so a return here means the
   * texture already matches [backgroundBitmap].
   */
  private fun uploadPendingBitmap() {
    val bitmap = backgroundBitmap ?: return
    if (composite.isBackgroundUploaded) return
    composite.uploadBackground(bitmap)
  }

  /**
   * Decodes a background, downscaled to at most [MAX_BACKGROUND_EDGE_PX] on its
   * longest edge.
   *
   * The downscale is not an optimisation, it is a correctness requirement. The
   * package's own bundled images go up to 8192x5464, which would decode to ~171 MB
   * of RGBA — an OOM on mid-range devices — and exceed the 4096px
   * `GL_MAX_TEXTURE_SIZE` that many GPUs report, making the texture upload fail
   * outright. Consumers can inject arbitrary images, so this cannot rely on assets
   * being well-sized.
   *
   * Supports the URI forms the JS side can produce: a `file://`/absolute path in
   * release, and an `http://` Metro URL in dev.
   *
   * Returns null on any failure — the renderer then passes the camera frame
   * through instead of showing a broken composite, and never logs.
   */
  private fun decode(uri: String): Bitmap? =
    try {
      // Two passes: measure, then decode subsampled. `inSampleSize` only honours
      // powers of two, so this lands at or below the cap, never above it.
      val bytes = readBytes(uri)
      val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
      BitmapFactory.decodeByteArray(bytes, 0, bytes.size, bounds)

      val longest = maxOf(bounds.outWidth, bounds.outHeight)
      var sampleSize = 1
      while (longest / sampleSize > MAX_BACKGROUND_EDGE_PX) {
        sampleSize *= 2
      }

      val options = BitmapFactory.Options().apply { inSampleSize = sampleSize }
      BitmapFactory.decodeByteArray(bytes, 0, bytes.size, options)
    } catch (_: Throwable) {
      null
    }

  private fun readBytes(uri: String): ByteArray =
    when {
      uri.startsWith("http://") || uri.startsWith("https://") ->
        URL(uri).openStream().use { it.readBytes() }
      uri.startsWith("file://") -> File(java.net.URI(uri)).readBytes()
      else -> File(uri).readBytes()
    }

  override fun connectSurface(surface: Surface, width: Int, height: Int) {
    viewportWidth = width
    viewportHeight = height
    egl.createWindowSurface(surface)
    // The old context's GL objects died with it; rebuild on the next frame.
    isGlReady = false
    composite.clearBackground()
  }

  override fun disconnectSurface() {
    egl.releaseSurface()
  }

  override fun dispose() {
    super.dispose()
    // A screen torn down mid-recording would otherwise leave an encoder running and
    // a half-muxed file nothing else knows the path of.
    synchronized(captureLock) {
      recorder?.release(deleteOutput = true)
      recorder = null
      pendingCapture = null
    }
    captureFinisher.shutdownNow()
    if (isGlReady) composite.release()
    isGlReady = false
    egl.release()
    backgroundBitmap?.recycle()
    backgroundBitmap = null
    segmenter.close()
  }
}
